### PR TITLE
Better perl

### PR DESCRIPTION
**Pull request by ialarmedalien**
Part I of some improvements I made a while back to the perl sdk app templates. I'm using the fully improved perl sdk app for the Jacobson / Cytoscape app (kb_cytoscape).

[Commit 1](https://github.com/kbase/kb_sdk/commit/420bc6b7146cdecec32585b42fd309a5aa9482f6) is almost entirely whitespace changes from running the code formatter `perltidy` over the codebase.

[Commit 2](https://github.com/kbase/kb_sdk/commit/48286f8c62e6186c4ddff7aefdfa84f1600efd00) has more simple changes, including:
- some more perltidy tidy up
- adding `use strict` and `use warnings` to all files so errors get reported
- use the correct form of class initialisation, which is `Class::Name->new`, not `new Class::Name`
- removed `$PATH` from the `$PERL5LIB` env var -- `$PATH` is for executables, `$PERL5LIB` is for libraries
- a few minor variable renamings

Note: the tests are marked as failed because this is from my own forked repo. They have passed on my repo: https://github.com/ialarmedalien/kb_sdk/tree/better_perl (note little green tick).

### CODE REVIEW

File: src/java/us/kbase/templates/module_dockerfile.vm.properties
Comment (ialarmedalien) [inline src/java/us/kbase/templates/module_dockerfile.vm.properties:6]: This tag has been deprecated

File: src/java/us/kbase/templates/module_dockerfile.vm.properties
Comment (ialarmedalien) [inline src/java/us/kbase/templates/module_dockerfile.vm.properties:42]: combine all directives into a single statement

File: src/java/us/kbase/templates/module_perl_impl.vm.properties
Comment (ialarmedalien) [inline src/java/us/kbase/templates/module_perl_impl.vm.properties:232]: Some of this is duplication of what's in the sample app, and unless you run `make`, the app that gets produced by default is just a bunch of random stanzas. I've integrated this and the sample app code together so that both the base app and the sample app are valid, sensible code without needing to run `make` / invoking the SDK.

File: src/java/us/kbase/templates/perl_client.vm.properties
Comment (ialarmedalien) [inline src/java/us/kbase/templates/perl_client.vm.properties:144]: moved the `# Authentication` line out of the method and into the header

File: src/java/us/kbase/templates/perl_client.vm.properties
Comment (ialarmedalien) [inline src/java/us/kbase/templates/perl_client.vm.properties:157]: `@_bad_arguments` doesn't need that underscore, especially since it gets reported to the user and is therefore not private!

**Comment by MrCreosote** [comment]
Sorry, I don't understand why the tests would fail in your repo but pass here...? Is it due to secrets not working on PRs?

File: lib/Bio/KBase/Log.pm
Comment (MrCreosote) [inline lib/Bio/KBase/Log.pm:266]: Having the conditional and the inner block have the same level of indentation makes it pretty hard to read quickly.

File: lib/Bio/KBase/Log.pm
Comment (MrCreosote) [inline lib/Bio/KBase/Log.pm:335]: I personally strongly dislike whitespace between brackets / parentheses and contents, but *shrug*

File: lib/Bio/KBase/Log.pm
Comment (MrCreosote) [inline lib/Bio/KBase/Log.pm:241]: putting the `elsif` on its own like makes it harder to read, as it's less clear that it's a continuation of the prior `if`.

File: lib/Bio/KBase/Log.pm
Comment (MrCreosote) [inline lib/Bio/KBase/Log.pm:248]: Same for `else`.

File: lib/Bio/KBase/Log.pm
Comment (MrCreosote) [reply lib/Bio/KBase/Log.pm:241]: Multiple places where this occurs in this file.

File: lib/Bio/KBase/Log.pm
Comment (MrCreosote) [reply lib/Bio/KBase/Log.pm:248]: Multiple places where this occurs in this file.

File: lib/Bio/KBase/Log.pm
Comment (MrCreosote) [inline lib/Bio/KBase/Log.pm:369]: Why would the die message go on a separate line?

Comment (MrCreosote) [reply lib/Bio/KBase/Log.pm:369]: This also occurs in multiple places.

File: src/java/us/kbase/templates/module_dockerfile.vm.properties
Comment (MrCreosote) [reply src/java/us/kbase/templates/module_dockerfile.vm.properties:6]: I would want @sychan to weigh in on this one since he created all the hooks/scripts etc that create and use these variables and he arguably best understands how they interact. Seems innocuous but worth double checking.

File: src/java/us/kbase/templates/module_dockerfile.vm.properties
Comment (MrCreosote) [inline src/java/us/kbase/templates/module_dockerfile.vm.properties:37]: Seems like this shouldn't be deleted, it's supposed to separate the part of the file the user can edit from the part they shouldn't

**Comment by ialarmedalien** [comment]
@MrCreosote yes, exactly.

File: src/java/us/kbase/templates/module_dockerfile.vm.properties
Comment (scanon) [reply src/java/us/kbase/templates/module_dockerfile.vm.properties:6]: We should replace it with LABEL maintainer="...."

File: src/java/us/kbase/templates/module_perl_impl.vm.properties
Comment (MrCreosote) [reply src/java/us/kbase/templates/module_perl_impl.vm.properties:232]: Thanks for this explanation, the changes threw me for a loop at first

File: src/java/us/kbase/templates/module_perl_impl.vm.properties
Comment (MrCreosote) [inline src/java/us/kbase/templates/module_perl_impl.vm.properties:99]: Should there be a newline before `Dumper $params;`?

File: src/java/us/kbase/templates/perl_client.vm.properties
Comment (MrCreosote) [reply src/java/us/kbase/templates/perl_client.vm.properties:144]: Sorry, I don't understand this change. Can you further explain what's going on here?

File: src/java/us/kbase/templates/perl_client.vm.properties
Comment (MrCreosote) [reply src/java/us/kbase/templates/perl_client.vm.properties:157]: Can you even access variables inside a function scope?

File: src/java/us/kbase/templates/perl_client.vm.properties
Comment (MrCreosote) [inline src/java/us/kbase/templates/perl_client.vm.properties:695]: Should this be indented?

File: src/java/us/kbase/templates/perl_client.vm.properties
Comment (MrCreosote) [inline src/java/us/kbase/templates/perl_client.vm.properties:717]: this isn't following the pattern of brace on `if` line

File: .gitignore
Comment (MrCreosote) [inline .gitignore:11]: What proportion of changes in this PR are covered by tests?

File: src/java/us/kbase/templates/module_dockerfile.vm.properties
Comment (ialarmedalien) [reply src/java/us/kbase/templates/module_dockerfile.vm.properties:6]: Done.

File: src/java/us/kbase/templates/perl_impl.vm.properties
Comment (MrCreosote) [inline src/java/us/kbase/templates/perl_impl.vm.properties:52]: Why the deletion? I'm not particularly familiar with perldoc

**Review by sychan — COMMENTED** [review]
This review seems to have stalled. Given that there are few people who have experience with the Perl functionality beyond AJ, I'm going to merge this by COB today unless I see a substantive criticism.

**Comment by MrCreosote** [comment]
I'm still working on it. There's just a lot of changes and I've had a lot of meetings the past couple days.

File: src/java/us/kbase/templates/perl_server.vm.properties
Comment (MrCreosote) [inline src/java/us/kbase/templates/perl_server.vm.properties:101]: I'm not a perl expert, but this looks like to me that the vars will be assigned the functions themselves, rather than the return value of the functions. If this is some sort of odd syntax where the functions get called even without the parens is there a reason to do it this way? It's very confusing to someone who doesn't know the trick.

File: src/java/us/kbase/templates/perl_server.vm.properties
Comment (MrCreosote) [inline src/java/us/kbase/templates/perl_server.vm.properties:174]: Having the same indentation for the conditional and the block is difficult to read

File: src/java/us/kbase/templates/perl_server.vm.properties
Comment (MrCreosote) [inline src/java/us/kbase/templates/perl_server.vm.properties:667]: This line is actually swapping the logic around a bit rather than just whitespace - is this tested somehow?

Comment (MrCreosote) [reply src/java/us/kbase/templates/perl_server.vm.properties:667]: Normally I wouldn't worry about it but the tests for kb_sdk don't run right now so I'm on the wary side

File: src/java/us/kbase/templates/perl_server.vm.properties
Comment (MrCreosote) [inline src/java/us/kbase/templates/perl_server.vm.properties:694]: The logic here is also shuffled around, so same question re testing

File: src/java/us/kbase/templates/perl_server.vm.properties
Comment (MrCreosote) [inline src/java/us/kbase/templates/perl_server.vm.properties:702]: logic shuffling

File: src/java/us/kbase/templates/perl_server.vm.properties
Comment (MrCreosote) [inline src/java/us/kbase/templates/perl_server.vm.properties:764]: Also logic shuffling here, but pretty minor logic shuffling

File: src/java/us/kbase/templates/module_dockerfile.vm.properties
Comment (MrCreosote) [reply src/java/us/kbase/templates/module_dockerfile.vm.properties:6]: That's interesting, review comments are added to the comment thread based on the day they're made (although the date still shows `now`) rather than being added to the end of the comment thread as one would expect

File: src/java/us/kbase/templates/perl_client.vm.properties
Comment (ialarmedalien) [reply src/java/us/kbase/templates/perl_client.vm.properties:144]: prior to change:
```
sub _check_job {
    my ($self, @args) = @_;
    # Authentication: ...
```

after change:
```
# Authentication
sub _check_job {
    my ($self, @args) = @_;
```

i.e. the comment isn't stuck in the middle of the method

File: src/java/us/kbase/templates/module_perl_impl.vm.properties
Comment (ialarmedalien) [reply src/java/us/kbase/templates/module_perl_impl.vm.properties:99]: No need.

File: src/java/us/kbase/templates/perl_client.vm.properties
Comment (ialarmedalien) [reply src/java/us/kbase/templates/perl_client.vm.properties:157]: You can't access `@bad_arguments` outside `_check_job`, no.

File: src/java/us/kbase/templates/perl_client.vm.properties
Comment (ialarmedalien) [reply src/java/us/kbase/templates/perl_client.vm.properties:695]: The indentation of this whole section is messed up -- there seems to be a lot of confusion on the part of the author on when new scopes need to be created... well, just general confusion about perl. I have addressed this in a future commit.

The line `#if( $enable_client_retry )` is a template directive, so its indentation does not matter too much as it won't appear in the output.

File: src/java/us/kbase/templates/perl_client.vm.properties
Comment (ialarmedalien) [reply src/java/us/kbase/templates/perl_client.vm.properties:717]: The perltidy linter didn't catch this, probably because it's got some template variables in the line (the `$codes_to_retry${empty_escaper}{$code}`). I have gone over a fair amount of this code by hand, so it's likely to be fixed in a future PR.

File: src/java/us/kbase/templates/perl_server.vm.properties
Comment (ialarmedalien) [reply src/java/us/kbase/templates/perl_server.vm.properties:101]: You don't have to use brackets when calling a function in perl; if there are no arguments, calling `$self->do_something` will do the same as `$self->do_something()`. There are a lot of functions that can be called with or without brackets -- e.g. `print`, `join`, `split`, `die`, etc., etc. Generally perl programmers will omit the brackets unless they are required to disambiguate an expression.

File: src/java/us/kbase/templates/perl_server.vm.properties
Comment (ialarmedalien) [reply src/java/us/kbase/templates/perl_server.vm.properties:174]: I just used the default perltidy setting.

File: src/java/us/kbase/templates/perl_server.vm.properties
Comment (ialarmedalien) [reply src/java/us/kbase/templates/perl_server.vm.properties:667]: In perl, you can put conditionals after an expression, e.g.

```perl
do_this( $args ) if $condition;
```
or
```perl
$var = $value unless defined $var;
```

It means you can write a lot of these types of conditional expressions in a single line, instead of having to put them in a `if (...) {  }` stanza.

File: src/java/us/kbase/templates/perl_server.vm.properties
Comment (ialarmedalien) [reply src/java/us/kbase/templates/perl_server.vm.properties:694]: Ditto explanation from above.

File: src/java/us/kbase/templates/perl_server.vm.properties
Comment (ialarmedalien) [reply src/java/us/kbase/templates/perl_server.vm.properties:764]: Not logic shuffling, just removing an unneeded `else`. If the `if` ends with a `return`, there's no need to put in an `else`.

File: src/java/us/kbase/templates/module_perl_impl.vm.properties
Comment (ialarmedalien) [reply src/java/us/kbase/templates/module_perl_impl.vm.properties:232]: It threw me for even more of a loop when I first generated a perl kbase app! At least now there's a sensible impl file returned after init-ing a repo.

File: lib/Bio/KBase/Log.pm
Comment (ialarmedalien) [reply lib/Bio/KBase/Log.pm:335]: I'm the opposite - I find it's easier to read code when things are more spread out.

File: lib/Bio/KBase/Log.pm
Comment (ialarmedalien) [reply lib/Bio/KBase/Log.pm:241]: this is the default perltidy setting

File: lib/Bio/KBase/Log.pm
Comment (ialarmedalien) [reply lib/Bio/KBase/Log.pm:369]: Maybe to keep the whole string intact? I don't know. I didn't look over the whitespace in every file in detail for this PR, this was just to impose standard formatting on all the perl files before looking over the code itself.

File: .gitignore
Comment (ialarmedalien) [reply .gitignore:11]: The vast majority of the changes are generated by perltidy, which is a popular and very commonly-used code formatter:

https://metacpan.org/pod/distribution/Perl-Tidy/bin/perltidy

This is the first in a series of updates to the perl templates that I have made and have been using on my own repo. I have added in tests for perl file compilation and rewritten the test script to use the perl test harness properly (I have also retrofitted those changes to a couple of the existing perl kbase apps -- e.g. RAST SDK; this means test failures will be reported, rather than just dying silently). This PR just applies some standardised formatting.

File: src/java/us/kbase/templates/module_dockerfile.vm.properties
Comment (ialarmedalien) [reply src/java/us/kbase/templates/module_dockerfile.vm.properties:6]: Docker docs on MAINTAINER tag: https://docs.docker.com/engine/reference/builder/#maintainer-deprecated

Comment (sychan) [reply src/java/us/kbase/templates/module_dockerfile.vm.properties:6]: The PR is correct, the MAINTAINER directive has been obsolete for a while and newer image should use LABEL MAINTAINER=

File: src/java/us/kbase/templates/perl_server.vm.properties
Comment (sychan) [reply src/java/us/kbase/templates/perl_server.vm.properties:667]: Yes, this is idiomatic Perl.

File: src/java/us/kbase/templates/perl_server.vm.properties
Comment (sychan) [reply src/java/us/kbase/templates/perl_server.vm.properties:694]: This just idiomatic perl.

File: lib/Bio/KBase/Log.pm
Comment (sychan) [reply lib/Bio/KBase/Log.pm:335]: This is a personal stylistic preference.

**Review by sychan — APPROVED** [review]
The majority of these changes are automatic formatting changes that perltidy applied, and while we may not personally like the styling, having a standard formatter is a huge step in the right direction.

The change to the Dockerfile was a legitimate question, but the change was actually correct, because the MAINTAINER directive became obsolete after the original dockerfile was generated.

Unless some more substantive issues come up, I think this PR is ok to merge.

**Pull request merged**
